resources/fetchersManager/data.go

```diff
@@ -30,48 +30,72 @@ import (
 // Data maintains a cache that is updated by Fetcher implementations registered
 // against it. It sends the cache to an output channel at the defined interval.
 type Data struct {
-	log      *logp.Logger
-	timeout  time.Duration
-	interval time.Duration
-	fetchers FetchersRegistry
-	// Wait for completion of fetcher's fetchSingle
-	wg   sync.WaitGroup
-	stop chan struct{}
+	log        *logp.Logger
+	timeout    time.Duration
+	interval   time.Duration
+	fetchers   FetchersRegistry
+	stop       chan struct{}
+	stopNotice chan time.Duration
 }
 
+type Stop func(context.Context, time.Duration)
+
 // NewData returns a new Data instance.
 // interval is the duration the manager wait between two consecutive cycles.
 // timeout is the maximum duration the manager wait for a single fetcher to return results.
 func NewData(log *logp.Logger, interval time.Duration, timeout time.Duration, fetchers FetchersRegistry) (*Data, error) {
 	return &Data{
-		log:      log,
-		timeout:  timeout,
-		interval: interval,
-		fetchers: fetchers,
-		wg:       sync.WaitGroup{},
-		stop:     make(chan struct{}),
+		log:        log,
+		timeout:    timeout,
+		interval:   interval,
+		fetchers:   fetchers,
+		stop:       make(chan struct{}),
+		stopNotice: make(chan time.Duration),
 	}, nil
 }
 
 // Run starts all configured fetchers to collect resources.
-func (d *Data) Run(ctx context.Context) error {
+func (d *Data) Run(ctx context.Context) Stop {
 	go d.fetchAndSleep(ctx)
-	return nil
+	called := false
+	return func(ctx context.Context, grace time.Duration) {
+		if called {
+			return
+		}
+		called = true
+		d.stopData(ctx, grace)
+	}
 }
 
 func (d *Data) fetchAndSleep(ctx context.Context) {
-	// Happens once in a lifetime of cloudbeat and then enters the loop
-	d.fetchIteration(ctx)
+	ctx, cancel := context.WithCancel(ctx)
+	// set immidiate exec for first time run
+	timer := time.NewTimer(0)
+	defer func() {
+		cancel()
+		timer.Stop()
+	}()
+	run := true
 	for {
 		select {
+		case grace := <-d.stopNotice:
```

Comment:
Is it possible to pass the main context to the stop function and cancel it after the grace period has ended, thereby achieving the same result?

Reply:
I am not sure I understand. The `stopNotice` channel comes to make sure that no new fetchers will be executed once `stop` is called. How passing the main context will achieve this?

Reply:
We can store the `cancelFunc` generated in `fetchAndSleep` and call it once `stop` is called.
```
	ctx, d.cancelFunc = context.WithCancel(ctx)

and then in `stop`

       d.cancelFunc()

```